xdsl/dialects/stablehlo.py

```diff
@@ -126,13 +128,69 @@ class SubtractOp(ElementwiseBinaryOperation):
     name = "stablehlo.subtract"
 
 
+@irdl_op_definition
+class TransposeOp(IRDLOperation):
+    """
+    Permutes the dimensions of `operand` tensor using `permutation` and produces a
+    `result` tensor. More formally, `result[result_index] = operand[operand_index]`
+    where `result_index[d] = operand_index[permutation[d]]`.
+
+    https://github.com/openxla/stablehlo/blob/main/docs/spec.md#transpose
+    """
+
+    name = "stablehlo.transpose"
+
+    ElementType = Annotated[Attribute, ConstraintVar("ElementType")]
+
+    operand = operand_def(TensorType[ElementType])
+    result = result_def(TensorType[ElementType])
+    permutation = attr_def(DenseArrayBase)
+
+    def __init__(
+        self, operand: SSAValue, permutation: DenseArrayBase, result_type: Attribute
+    ):
+        super().__init__(
+            operands=(operand,),
+            result_types=(result_type,),
+            attributes={"permutation": permutation},
+        )
+
+    def get_permutation(self) -> tuple[int, ...]:
+        return cast(tuple[int, ...], self.permutation.as_tuple())
+
+    def verify_(self) -> None:
+        # Operand and result types are checked before the custom `verify_`
+        o_type = cast(TensorType[Attribute], self.operand.type)
+        r_type = cast(TensorType[Attribute], self.result.type)
+
+        o_shape = o_type.get_shape()
+        r_shape = r_type.get_shape()
+
+        # TODO: C1 - quantization
```

Comment:
What do `C1` `C2` and `C3` mean here?

Reply:
They're the constraint names in the original stablehlo documentation

Reply:
Would it be worth explaining this/linking in the comment?

Reply:
Actually I just removed them, they're not even consistently numbered throughout the document